### PR TITLE
[24.0] Fix authentication error for anonymous users querying jobs

### LINKED ISSUE #18331: GET /api/jobs fails with 500 Internal Server Error when not logged in
**Describe the bug**
When visiting https://usegalaxy.eu/api/jobs when not logged in, it fails with a HTTP error 500 (Internal Server Error) instead of 403. Same on usegalaxy.org .

**Galaxy Version and/or server at which you observed the bug**
Galaxy Version: 24.0

**To Reproduce**
1. Visit https://usegalaxy.eu/api/jobs when not logged in (e.g. on a private browser window), or run `requests.get("https://usegalaxy.eu/api/jobs")` on an Python shell
2. Error 500 (Internal Server Error) is returned

**Expected behavior**
A 403 error, not a server crash.

### PR DESCRIPTION
**Pull request by davelopez**
Fixes #18331

I added the check at the API level (to fail fast) and at the manager level in case it gets called from somewhere else.

## How to test the changes?
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

### CODE REVIEW

File: lib/galaxy/managers/jobs.py [hunk 1]
```diff
@@ -111,6 +112,8 @@ def index_query(self, trans, payload: JobIndexQueryPayload) -> sqlalchemy.engine
         Otherwise this will only return the user's jobs or all jobs if the requesting
         user is acting as an admin.
         """
+        if trans.user is None:
```
Comment (mvdbeek) [inline lib/galaxy/managers/jobs.py]: Can we instead query the jobs using the session id or current session history ? I think there's no reason anon user can't see their jobs

Comment (davelopez) [reply lib/galaxy/managers/jobs.py]: Sure! I forgot there was a way to differentiate jobs from other anonymous users. I'll give it a try :+1:

Comment (davelopez) [reply lib/galaxy/managers/jobs.py]: I tried to write an API or selenium test but I couldn't figure out how to use the session. I manually tested and seems to work though.

File: lib/galaxy/managers/jobs.py [hunk 2]
```diff
@@ -105,7 +106,9 @@ def __init__(self, app: StructuredApp):
         self.app = app
         self.dataset_manager = DatasetManager(app)
 
-    def index_query(self, trans, payload: JobIndexQueryPayload) -> sqlalchemy.engine.Result:
+    def index_query(
+        self, trans: ProvidesUserContext, payload: JobIndexQueryPayload
+    ) -> typing.Optional[sqlalchemy.engine.Result]:
```
Comment (nsoranzo) [inline lib/galaxy/managers/jobs.py]: I'd add `from typing import Optional` and use just `Optional` here, in preparation for more type annotations.

File: lib/galaxy/managers/jobs.py [hunk 3]
```diff
@@ -121,6 +127,13 @@ def index_query(self, trans, payload: JobIndexQueryPayload) -> sqlalchemy.engine
         search = payload.search
         order_by = payload.order_by
 
+        if trans.user is None:
+            # If the user is anonymous we can only return jobs for the current session history
+            if trans.galaxy_session and trans.galaxy_session.current_history_id:
+                history_id = trans.galaxy_session.current_history_id
+            else:
+                return None
```
Comment (nsoranzo) [inline lib/galaxy/managers/jobs.py:135]: I'm wondering if we should raise an `AuthenticationRequired` exception here instead, so the error is not hidden from the user.

Comment (mvdbeek) [reply lib/galaxy/managers/jobs.py:135]: This is not an error, if a session doesn't have a history it also doesn't have any jobs.

Comment (nsoranzo) [reply lib/galaxy/managers/jobs.py:135]: But None is returned also is `trans.galaxy_session` is None.

Comment (mvdbeek) [reply lib/galaxy/managers/jobs.py:135]: Which is fine and not a error ?

Comment (mvdbeek) [reply lib/galaxy/managers/jobs.py:135]: You could as well query some public thing, like jobs belonging to a history, raising an exception here feels very unpredictable.

Comment (nsoranzo) [reply lib/galaxy/managers/jobs.py:135]: Then I don't understand the changes in lines 131-133 above: if we want a session-less user to be able to view public history jobs, we should overwrite `history_id` only when it's `None` .

Comment (mvdbeek) [reply lib/galaxy/managers/jobs.py:135]: That is a good point, we might want to instead filter on the job itself using the session id.

Comment (nsoranzo) [reply lib/galaxy/managers/jobs.py:135]: If `history_id` is passed in the API payload, the job's history accessibility is checked on the jobs returned by this method at https://github.com/galaxyproject/galaxy/blob/7caee0b7511152937e51229c28cdb778824d122d/lib/galaxy/webapps/galaxy/services/jobs.py#L89

Comment (davelopez) [reply lib/galaxy/managers/jobs.py:135]: Thanks for the catch! I'll follow up with the proposed change :+1:

**Pull request merged**